compiler/src/dotty/tools/dotc/typer/PrepareTransparent.scala

```diff
@@ -0,0 +1,411 @@
+            ctx.warning("implicit reference $tree is dropped at inline site because it refers to local symbol(s)", tree.pos)
+          else {
+            implicitRefTypes += tree.tpe
+            implicitRefs += tree
+          }
+        case _ =>
+      }
+
+      def registerAccessor(tree: Tree) = {
+        inlining.println(i"accessor: $tree at ${tree.pos}")
+        accessorAtPos(tree.pos.toSynthetic) = tree.symbol
+      }
+
+      def traverse(tree: Tree)(implicit ctx: Context): Unit = {
+        tree match {
+          case Ident(nme.WILDCARD) =>
+          case _: Ident | _: This =>
+            //println(i"leaf: $tree at ${tree.pos}")
+            if (tree.symbol.exists && !isLocal(tree.symbol, inlineMethod)) {
+              if (ctx.debug) inlining.println(i"type at $tree @ ${tree.pos.toSynthetic} = ${tree.tpe}")
+              typeAtPos(tree.pos.toSynthetic) = tree.tpe
+            }
+          case _: Select =>
+            tree.symbol.name match {
+              case InlineAccessorName(UniqueInlineName(_, _)) => return // was already recorded in Apply
+              case InlineAccessorName(_) => registerAccessor(tree)
+              case _ =>
+            }
+          case Apply(_: RefTree | _: TypeApply, receiver :: Nil) =>
+            tree.symbol.name match {
+              case InlineAccessorName(UniqueInlineName(_, _)) => registerAccessor(tree)
+              case _ =>
+            }
+          case _ =>
+        }
+        registerIfContextualImplicit(tree)
+        traverseChildren(tree)
+      }
+    }
+    referenced.traverse(typed)
+
+    // The untyped tree transform that applies the tweaks
+    object addRefs extends untpd.UntypedTreeMap {
+      override def transform(tree: untpd.Tree)(implicit ctx: Context): untpd.Tree = {
+
+        def adjustLeaf(tree: untpd.Tree): untpd.Tree = referenced.typeAtPos.get(tree.pos.toSynthetic) match {
+          case Some(tpe) => untpd.TypedSplice(tree.withType(tpe))
+          case none => tree
+        }
+
+        def adjustForAccessor(ref: untpd.RefTree) =
+          referenced.accessorAtPos.get(ref.pos.toSynthetic) match {
+            case Some(acc) =>
+              def accessorRef = untpd.TypedSplice(tpd.ref(acc))
+              acc.name match {
+                case InlineAccessorName(UniqueInlineName(_, _)) =>
+                  // In this case we are seeing a pair like this:
+                  //   untyped                typed
+                  //   t.x                    inline$x(t)
+                  // Drop the selection, since it is part of the accessor
+                  val Select(qual, _) = ref
+                  untpd.Apply(accessorRef, qual :: Nil)
+                case _ =>
+                  accessorRef
+              }
+            case none => ref
+          }
+
+        def adjustQualifier(tree: untpd.Tree): untpd.Tree = tree match {
+          case tree @ Ident(name1) =>
+            referenced.typeAtPos.get(tree.pos.startPos) match {
```

Comment:
I don't think that's good enough, if I have an identifier `foo` it might get typed with multiple qualifiers `a.b.foo` each of those will have the same zero-extent position `tree.pos.startPos`.

Reply:
If it is a non-trivial qualifier the tree gets replaced by a Select node, no?